cpp/include/cudf/interop.hpp

```diff
@@ -130,6 +140,298 @@ using unique_schema_t = std::unique_ptr<ArrowSchema, void (*)(ArrowSchema*)>;
+ */
+std::vector<cudf::column_metadata> get_table_metadata(cudf::table_view const& input);
+
+/**
+ * @brief A standard interchange medium for ArrowDeviceArray data in cudf.
+ *
+ * This class provides a way to work with ArrowDeviceArray data in cudf without
+ * sacrificing the APIs expected of a cudf column. On the other end, it
+ * provides the shared lifetime management expected by arrow consumers rather
+ * than the single-owner mechanism of cudf::column.
+ */
+class arrow_column {
+ public:
+  /**
+   * @brief Construct a new arrow column object
+   *
+   * The input array will be moved into the arrow_column, so it is no longer
+   * suitable for use afterwards.
+   *
+   * @param schema Arrow schema for the column
+   * @param input ArrowDeviceArray data for the column
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used for any allocations during conversion
+   */
+  arrow_column(ArrowSchema const* schema,
+               ArrowDeviceArray* input,
+               rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+               rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
+
+  /**
+   * @brief Construct a new arrow column object
+   *
+   * The input array will be released, so it is no longer suitable for use
+   * afterwards.
+   *
+   * @param schema Arrow schema for the column
+   * @param input ArrowArray data for the column
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used for any allocations during conversion
+   */
+  arrow_column(ArrowSchema const* schema,
+               ArrowArray* input,
+               rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+               rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
+
+  /**
+   * @brief Construct a new arrow column object
+   *
+   * The input column will be moved into the arrow_column, so it is no longer
+   * suitable for use afterwards.
+   *
+   * @param input cudf column to convert to arrow
+   * @param metadata Column metadata for the column
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used for any allocations during conversion
+   */
+  arrow_column(cudf::column&& input,
+               column_metadata const& metadata,
+               rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+               rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
+
+  /**
+   * @brief Convert the column to an ArrowSchema
+   *
+   * @param output ArrowSchema to populate with the column's schema
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used for any allocations during conversion
+   */
+  void to_arrow_schema(ArrowSchema* output,
```

Comment:
can we specify the lifetime of the `ArrowSchema` populated here? Does the output *own* it meaning that we performed a deep copy from the one that was passed in during construction?

Reply:
Yes, we are performing a deep copy. My original implementation did not do this and instead pointed non-value-type fields (like the name and format pointers to strings) back into the input values, but it also wasn't clear to me from the spec whether there is any standard for shallow copying a schema or what the expectations of a consumer would be for that. My presumption is that the schema is cheap to copy. We can revisit that if you think it is useful to change those semantics, WDYT?

Reply:
In general, you're correct that the Schema is cheap to copy. It only gets expensive when you start having very wide schemas or highly nested schemas with a ton of children. You aren't going to want to be deep copying schemas frequently (purely because they end up being a lot of small allocations) but it shouldn't be a problem in 95% of cases I'd wager.

Reply:
What is the alternative? Is it acceptable to return shallow copies that only free the parts that they own themselves? I had originally written a shallow copy function similar to `copy_array` that we could use instead, but unlike with arrays it's not as obvious what should be shallow copied. Anything that is a pointer?

I'm also fine punting on this and leaving a deep copy until we notice the perf impact in a practical workflow.

Reply:
I say we should punt on this and leave it as a deep copy until you notice a perf impact in practical workflows. Like I said, this won't generally be an issue or bottleneck unless you have a very wide or deeply nested schema.

Reply:
Cool, added notes on the lifetime to the docs.